container/state.go

```diff
@@ -160,64 +165,78 @@ func IsValidStateString(s string) bool {
 	return true
 }
 
-func wait(waitChan <-chan struct{}, timeout time.Duration) error {
-	if timeout < 0 {
-		<-waitChan
-		return nil
-	}
-	select {
-	case <-time.After(timeout):
-		return fmt.Errorf("Timed out: %v", timeout)
-	case <-waitChan:
-		return nil
-	}
-}
+// WaitCondition is an enum type for different states to wait for.
+type WaitCondition int
+
+// Possible WaitCondition Values.
+//
+// WaitConditionNotRunning (default) is used to wait for any of the non-running
+// states: "created", "exited", "dead", "removing", or "removed".
+//
+// WaitConditionNextExit is used to wait for the next time the state changes
+// to a non-running state. If the state is currently "created" or "exited",
+// this would cause Wait() to block until either the container runs and exits
+// or is removed.
+//
+// WaitConditionRemoved is used to wait for the container to be removed.
+const (
+	WaitConditionNotRunning WaitCondition = iota
+	WaitConditionNextExit
+	WaitConditionRemoved
+)
+
+// Wait waits until the continer is in a certain state indicated by the given
+// condition. A context can be used for cancelling the request or controlling
+// timeouts. Wait must be called without holding the state lock. Returns a
+// channel which can be used to receive the result. If the container exited on
+// its own, the result's Err() method wil be nil and its ExitCode() method will
+// return the conatiners exit code, otherwise, the results Err() method will
+// return an error indicating why the wait operation failed.
+func (s *State) Wait(ctx context.Context, condition WaitCondition) <-chan *StateStatus {
```

Comment:
So if it didn't return a channel then it would have to block? Wouldn't you still want to pass in a context to cancel that?

The reason this returns a channel (where `WaitStop` did not before) is because this call is usually wrapped by a call to `something.ContainerWait(ctx, containerID)` which returns the pair `(chan Status, error)`. If there is an error before the wait can begin then the error value will be non-nil. Once the wait begins though, the call no longer blocks and you get a channel back and the caller *knows that there is a backend goroutine blocking on the container to exit or be removed*. This allows the caller do do other things, synchronously, before pulling the value out of the channel, such as:

- create a container
- get channel to wait for that container's next exit
- start the container
- block on receiving container's exit status

If this did not return a channel than there would be race conditions in this use case:

- create container
- spawn goroutine to wait for container's next exit
- start container
- block on waiting for container's exit

In this scenario it's possible that you start the container and it exits before that previously spawned goroutine actually gets a chance to wait for the container and would then wait for the exit which would never happen again.

In the case of the API handler for `ContainerWait`, the returned `(chan Status, error)` values allow us to always return response headers immediately once the wait begins (or return `No Such Container` in the case of an error). That allows for the *"okay, the backend is now waiting, I'll send you the response once I have it"* flow to work through the API.